lib/Frontend/Frontend.cpp

```diff
@@ -261,233 +262,300 @@ ModuleDecl *CompilerInstance::getMainModule() {
-    if (importedHeaderModule)
-      additionalImports.push_back({ { /*accessPath=*/{}, importedHeaderModule },
-                                    SourceFile::ImportFlags::Exported });
-    if (!importModules.empty()) {
-      for (auto &importModule : importModules) {
-        additionalImports.push_back({ { /*accessPath=*/{}, importModule }, {} });
-      }
-    }
+  PersistentParserState PersistentState;
 
-    SF->addImports(additionalImports);
-  };
+  ensureMainFileComesFirst(implicitModuleImportKind,
+                           objCModuleUnderlyingMixedFramework,
+                           importedHeaderModule, importModules);
 
-  if (Kind == InputFileKind::IFK_Swift_REPL) {
-    auto *SingleInputFile =
-      new (*Context) SourceFile(*MainModule, Invocation.getSourceFileKind(),
-                                None, modImpKind);
-    MainModule->addFile(*SingleInputFile);
-    addAdditionalInitialImports(SingleInputFile);
+  bool hadLoadError = parsePartialModulesAndLibraryFiles(
+      implicitModuleImportKind, objCModuleUnderlyingMixedFramework,
+      importedHeaderModule, importModules, PersistentState, DelayedCB.get());
+  if (hadLoadError)
     return;
-  }
 
-  std::unique_ptr<DelayedParsingCallbacks> DelayedCB;
-  if (Invocation.isCodeCompletion()) {
-    DelayedCB.reset(
-        new CodeCompleteDelayedCallbacks(SourceMgr.getCodeCompletionLoc()));
-  } else if (Invocation.isDelayedFunctionBodyParsing()) {
-    DelayedCB.reset(new AlwaysDelayedCallbacks);
-  }
+  parseMainAndTypeCheckTopLevelFiles(PersistentState, DelayedCB.get());
+}
 
-  PersistentParserState PersistentState;
+// Return true if should continue, i.e. no error
+bool CompilerInstance::loadStdlibAndMaybeSwiftOnoneSupport() {
+  SharedTimer timer("performSema-loadStdlibAndMaybeSwiftOnoneSupport");
+  ModuleDecl *M = Context->getStdlibModule(true);
 
-  // Make sure the main file is the first file in the module. This may only be
-  // a source file, or it may be a SIL file, which requires pumping the parser.
-  // We parse it last, though, to make sure that it can use decls from other
-  // files in the module.
-  if (MainBufferID != NO_SUCH_BUFFER) {
-    assert(Kind == InputFileKind::IFK_Swift || Kind == InputFileKind::IFK_SIL);
+  if (!M) {
+    Diagnostics.diagnose(SourceLoc(), diag::error_stdlib_not_found,
+                         Invocation.getTargetTriple());
+    return false;
+  }
 
-    if (Kind == InputFileKind::IFK_Swift)
-      SourceMgr.setHashbangBufferID(MainBufferID);
+  // If we failed to load, we should have already diagnosed
+  if (M->failedToLoad()) {
+    assert(Diagnostics.hadAnyError() &&
+           "Module failed to load but nothing was diagnosed?");
+    return false;
+  }
+  if (shouldImplicityImportSwiftOnoneSupportModule(
+          Invocation.getSILOptions().Optimization,
+          Invocation.getFrontendOptions().RequestedAction)) {
+    Invocation.getFrontendOptions().ImplicitImportModuleNames.push_back(
```

Comment:
This doesn't actually load SwiftOnoneSupport itself, so it doesn't belong here.

Reply:
Could you please clarify? AFAICT, the code in question asks if it should implicitly import the module and if so, adds it to a list of modules to be imported. Did you want something like
'shouldArrangeForSwiftOnOneSupportToBeImportedLater'? What am I missing here? Tx.

Reply:
There's a difference between that and what it's doing for the standard library. We check if the standard library exists up front, but not SwiftOnoneSupport, and conversely we *don't* decide whether to implicitly import the standard library here, but we decide for SwiftOnoneSupport.

Reply:
I get it, I think. I'll separate this code out from where I had put it.

Reply:
As a side effect of separation, I made a minor functional change: before if stdlib failed to load, the Onone module was still added to the ImplicitImportModuleNames set. Now, if stdlib fails to load, the Onone module is not added.